### PR TITLE
[inductor] Improve sort kernel perf

### LINKED ISSUE #129507: Investigate alternatives to remove mask from triton sort kernel
Follow up to https://github.com/pytorch/pytorch/pull/128458#discussion_r1651871500

Investigate:
1. Using infinity/max-value + stable sort instead of mask
2. Using a negative index value as the mask to reuse the same reduction for both index and mask

cc @ezyang @anijain2305 @chauhang @penguinwu @voznesenskym @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire

### PR DESCRIPTION
**Pull request by peterbell10**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #131719

Closes #129507

This makes two changes to the sort kernel:
1. Use int16 for the indices since we only operate on small dims anyway
2. Instead of passing an explicit mask, we pass the rnumel and imply the
   mask from that which saves an additional reduction in the sort
   kernel's inner loop.

In my benchmarks, this gives enough of a perf improvement to bump up the
max rblock to 512.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang

### CODE REVIEW

File: torch/_inductor/lowering.py [hunk 1]
```diff
@@ -5671,8 +5671,11 @@ def sort_stable(x, *, stable=None, dim=-1, descending=False):
         return clone(x), _full(0, device, torch.int64, shape)
 
     dim_size = shape[dim] if len(shape) else 1
+    if not V.graph.sizevars.statically_known_lt(dim_size, torch.iinfo(torch.int16).max):
+        return sort_fallback(x, stable=stable, dim=dim, descending=descending)
+
     indices = iota(
-        dim_size, start=0, step=1, dtype=torch.int64, device=device, requires_grad=False
+        dim_size, start=0, step=1, dtype=torch.int16, device=device, requires_grad=False
```
Comment (eellison) [inline torch/_inductor/lowering.py:5678]: How important was int16 vs int32 for perf ? could remove above limitation

**Review by eellison — APPROVED** [review]
sweet!

File: torch/_inductor/lowering.py (hunk 1 above)
Comment (peterbell10) [reply torch/_inductor/lowering.py:5678]: It's only a few percent faster, but considering we fallback for anything over 9 bits anyway we might as well take it.
